bits/spf_discovery/internetnl_spf_parser.py

```diff
@@ -0,0 +1,155 @@
+macro_expand = (
+    (CaselessLiteral("%{") + macro_letter + Optional(transformers) + ZeroOrMore(delimiter) + CaselessLiteral("}"))
+    | CaselessLiteral("%%")
+    | CaselessLiteral("%_")
+    | CaselessLiteral("%-")
+)
+
+
+def _check_toplabel(tokens):
+    if tokens[0][-1] == "-":
+        raise ParseException("Top level ending in '-'")
+    return None
+
+
+toplabel = (
+    (Optional(Word(nums)) + Word(alphas, exact=1) + Optional(Word(alphanums)))
+    | (Word(alphanums) + CaselessLiteral("-") + Optional(Word(alphanums + "-")))
+).setParseAction(_check_toplabel)
+
+
+def _check_domain_end(tokens):
+    """
+    domain_end = (
+        (CaselessLiteral('.') + toplabel + Optional(CaselessLiteral('.')))
+        | macro_expand)
+
+    """
+    domain_name = tokens[0]
+    if domain_name[-1] == ".":
+        domain_end = domain_name.split(".")[-2]
+    else:
+        domain_end = domain_name.split(".")[-1]
+    try:
+        toplabel.parseString(domain_end)
+    except ParseException:
+        macro_expand.parseString(domain_end)
+    return None
+
+
+macro_string = Combine(ZeroOrMore(macro_expand | macro_literal))
+
+domain_spec = macro_string.setParseAction(_check_domain_end)
+
+ip4_network = Regex(
+    r"((25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}" "(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])"
+)
+
+ip6_cidr_length = CaselessLiteral("/") + Regex("(12[0-8]|1[01][0-9]|[1-9][0-9]|[0-9])")
+ip4_cidr_length = CaselessLiteral("/") + Regex("(3[0-2]|[12][0-9]|[0-9])")
+dual_cidr_length = Optional(ip4_cidr_length) + Optional(CaselessLiteral("/") + ip6_cidr_length)
+
+unknown_modifier = Combine(name + CaselessLiteral("=") + macro_string)
+explanation = Combine(CaselessLiteral("exp=") + domain_spec)
+redirect = Combine(CaselessLiteral("redirect=") + domain_spec)
+modifier = redirect | explanation | unknown_modifier
+
+qualifier = Word("+-?~", exact=1)
+exists = Combine(Optional(qualifier) + CaselessLiteral("exists:") + domain_spec)
+ip6 = Combine(Optional(qualifier) + CaselessLiteral("ip6:") + Regex("[^ ]*").setParseAction(_parse_ipv6))
+ip4 = Combine(Optional(qualifier) + CaselessLiteral("ip4:") + ip4_network + Optional(ip4_cidr_length))
+ptr = Combine(Optional(qualifier) + CaselessLiteral("ptr") + Optional(CaselessLiteral(":") + domain_spec))
+mx = Combine(
+    Optional(qualifier)
+    + CaselessLiteral("mx")
+    + Optional(CaselessLiteral(":") + domain_spec)
+    + Optional(dual_cidr_length)
+)
+a = Combine(
+    Optional(qualifier)
+    + CaselessLiteral("a")
+    + Optional(CaselessLiteral(":") + domain_spec)
+    + Optional(dual_cidr_length)
+)
+include = Combine(Optional(qualifier) + CaselessLiteral("include:") + domain_spec)
+all = Combine(Optional(qualifier) + CaselessLiteral("all"))
+
+mechanism = all | include | a | mx | ptr | ip4 | ip6 | exists
+directive = mechanism
+terms = ZeroOrMore(OneOrMore(SP) + (directive | modifier))
+
+version = CaselessLiteral("v=spf1").setResultsName("spf_version")
```

Comment:
A lot of methods you're using here in this current pyparsing version is deprecated, please change them to the appropriate alternative (e.g. the `setResultName`, `parseString` calls are renamed or deprecated due to other reasons)